A62-pick-first.md

```diff
@@ -0,0 +1,189 @@
+
+Because `pick_first` sends all requests to the same address, it is often used
+for L4 load balancing by randomizing the order of the addresses used by each
+client.  In general, gRPC expects address ordering to be determined as part of
+name resolution, not by the LB policy. For example, DNS servers may randomize
+the order of addresses when there are multiple A/AAAA records, and the DNS
+resolver in gRPC is expected to perform [RFC-6724][6724] address sorting.
+However, there are some cases where DNS cannot randomize the address order,
+either because the DNS server does not support that functionality or because it
+is defeated by client-side DNS caching. To address such cases, it is desirable
+to add a client-side mechanism for randomly shuffling the order of the
+addresses.
+
+[6274]: https://www.rfc-editor.org/rfc/rfc6724.html
+
+### `pick_first` via xDS
+
+There are cases where it is desirable to perform L4 load balancing using
+`pick_first` when getting addresses via xDS instead of DNS. As a result, we need
+a way to configure use of this LB policy via xDS.
+
+Note that client-side address shuffling may be equally desirable in this case,
+since the xDS server may send the same EDS resource (with the same endpoints in
+the same order) to all clients.
+
+### Related Proposals:
+
+* [gRFC A37: xDS Aggregate and Logical DNS Clusters][A37]
+* [gRFC A52: gRPC xDS Custom Load Balancer Configuration][A52]
+* [gRFC A56: `priority_experimental` LB policy][A56]
+
+[A37]: A37-xds-aggregate-and-logical-dns-clusters.md
+[A52]: A52-xds-custom-lb-policies.md
+[A56]: A56-priority-lb-policy.md
+
+## Proposal
+
+Specific changes are described in their own subsections below.
+
+### Use sticky-TF by default
+
+Using sticky-TF by default in all `pick_first` implementations would enable us
+to overcome the shortcomings described [here](#sticky-transient-failure). This
+would involve making the following changes to `pick_first` implementations, once
+connections to all addresses fail:
+- Report `TRANSIENT_FAILURE` as the connectivity state.
+- Attempt to reconnect to the addresses indefinitely until a connection succeeds
+  (at which point, they should report `READY`), or there is no RPC activity on
+  the channel for the specified `IDLE_TIMEOUT`.
+
+All gRPC implementations should implement `IDLE_TIMEOUT` and have it enabled by
+default. A default value of 30 minutes is recommended.
+
+### Enable random shuffling of address list
+
+At the time of this writing, `pick_first` implementations do not expect any
+configuration to be passed to it. As part of this design, we will add a field to
+its configuration that would enable it to randomly shuffle the order of the
+addresses it receives before it starts to attempt to connect.
+
+```
+{
+    // If set to true, instructs the LB policy to shuffle the order of the
+    // list of addresses received from the name resolver before attempting to
+    // connect to them.
+    "shuffleAddressList": boolean
```

Comment:
Please also send a PR to add this new field to the [service config proto](https://github.com/grpc/grpc-proto/blob/26635376b3f47a11126a0f4b4b5b6de7fe5a074a/grpc/service_config/service_config.proto#L194).

Reply:
Will do. Thanks.